### PR TITLE
Fix crash when `initWithValue:` nested object w/ uninitialized schema

### PR DESCRIPTION
**Pull request by JadenGeller**
Fixes https://github.com/realm/realm-cocoa/issues/3690. @jpsim @tgoyne

### CODE REVIEW

**Comment by JadenGeller** [comment]
Will this break for Swift objects (with mangled names)? I guess I ought to write tests to find out...

**Comment by JadenGeller** [comment]
I think fixing this in Swift also will require changing our property parsing code to also store the property `Class` instead of just the class name. Any reason we shouldn't do this?

**Comment by tgoyne** [comment]
There's already some [similar tests in SchemaTests.mm](https://github.com/realm/realm-cocoa/blob/master/Realm/Tests/SchemaTests.mm#L673) which take the approach of running the test in a pristine child process rather than trying to modify a bunch of global state and then restore it afterwards.

File: Realm/Tests/PreinitializationTests.mm [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+- (void)testDisallowRealmConstruction {
+    RLMAssertThrowsWithReasonMatching(RLMRealm.defaultRealm, @"Illegal to initialize a Realm in this test case.");
+}
+
+- (void)testCreateUnmanagedObject {
+    XCTAssertNoThrow([[IntObject alloc] initWithValue:@[@0]]);
+}
+
+- (void)testCreateUnmanagedObjectWithNestedObject {
+    id value = @[@0, @[@[@0]]]; // on separate line due to parsing bug
```
Comment (tgoyne) [inline Realm/Tests/PreinitializationTests.mm]: C preprocessor only groups on parentheses, so any other comma-separated list used within a macro needs to be surrounded with them.

**Comment by JadenGeller** [comment]
@tgoyne It's unfortunate that this makes debugging with breakpoints much more difficult though.

File: Realm/RLMSchema_Private.h [hunk 2]
```diff
@@ -51,4 +51,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @end
 
+// Used in `PreinitializationTests` to reset the schema before each test case
+void RLMUnsafeResetSchema();
```
Comment (bdash) [inline Realm/RLMSchema_Private.h]: Can this be removed now?

Comment (JadenGeller) [inline Realm/RLMSchema_Private.h]: Yes.

**Comment by jpsim** [comment]
This is breaking `RealmSwift_Tests.SchemaTests.testNoSchemaForUnmanagedObjectClasses`:

> XCTAssertNil failed: throwing "+[RLMObjectBase ignoredProperties]: unrecognized selector sent to class 0x10a33acb8" -

**Comment by JadenGeller** [comment]
Looks like this is currently passing Swift tests, so I think it's done. 👌

File: Realm/Tests/Swift/SwiftSchemaTests.swift [hunk 3]
```diff
@@ -106,6 +106,16 @@ class SwiftSchemaTests: RLMMultiProcessTestCase {
         let _ = SwiftCompanyObject()
     }
 
+    func testCreateUnmanagedObjectWhichCreatesAnotherClassViaInitWithValueDuringSchemaInit() {
```
Comment (jpsim) [inline Realm/Tests/Swift/SwiftSchemaTests.swift]: missing Swift 2.x version

**Comment by jpsim** [comment]
Missing a changelog entry and a Swift 2.x version of the test. Otherwise 👍

**Comment by JadenGeller** [comment]
One thing that's bugging me about this: The contents of the `objectSchemas` array is NOT the same as the schemas returned by `schemaForClassName:`. I think the solution to this is that accessing `objectSchemas` should FULLY initialize the object schema. (Also, I think that it's incorrect behavior that `schemaForClassName:` looks up unknown object schemas for any arbitrary `RLMSchema` instance—it should only do this for the shared schema).

**Pull request merged**